### DISCUSSION TITLE
Enforce validator on creation only

### DISCUSSION THREAD

**Discussion by domainzero** [netbox-community/netbox#10590] [Q&A] [answered]
There's probably a better way to do this, but I might be too blind to see it.
  
I have a custom validator to enforce uniqueness on device names with tenants, but the validator prevents me from editing existing objects as the name already exists. Is there a way to only enforce this validator on device _creation_ versus _update_? 

My validator looks like this:

```python
class DeviceNameUnique(CustomValidator):
    def validate(self, device):
        existing_devices = Device.objects.filter(name__iexact=device.name).exists()
        if existing_devices:
            self.fail("Device name in use.", field="name")
```

**Comment by kkthxbye-code** [comment, marked as answer]
You should be able to do:

`if not device.pk:`

device.pk will only be None if it's a creation.

**Reply by jeremystretch** [reply]
Alternatively:

```python
existing_devices = Device.objects.exclude(pk=device.pk).filter(name__iexact=device.name).exists()
```

**Reply by domainzero** [reply]
Thanks!

**Reply by jonasbela** [reply]
I'm trying to validate in the same way my serial numbers but i don't seem to get it working.
https://github.com/netbox-community/netbox/issues/9187

This was the original proposed code, but when i edit a device i run into the same issue as mentioned above:
from extras.validators import CustomValidator

```python
class MyValidator(CustomValidator):

    def validate(self, instance):
        from dcim.models import Device

        if instance.serial and Device.objects.filter(serial=instance.serial).exists():
            self.fail("Duplicate serial", field='serial')
```

I tried
```python
def validate(self, instance):
    if instance.serial and Device.objects.exclude(pk=device.pk).filter(serial=instance.serial).exists():
```

and
```python
def validate(self, device):
    existing_devices = Device.objects.filter(name__iexact=device.serial).exists()
```

and

```python
def validate(self, device):
    existing_devices = Device.objects.exclude(pk=device.pk).filter(name__iexact=device.serial).exists()
```

But none will let my netbox start
I have to add that i am a complete beginner at programming.

**Reply by kkthxbye-code** [reply]
> But none will let my netbox start

You have to post the error, no reason to make people guess the problem.

In your three examples it doesn't look like you are importing the Device model.

**Reply by jonasbela** [reply]
The problem was using tab instead of spaces - Thanks for pointing out to look into the log!

Jan 02 11:54:52 server2 gunicorn[45000]:   File "/opt/netbox/netbox/netbox/configuration.py", line 245
Jan 02 11:54:52 server2 gunicorn[45000]:             from dcim.models import Device
Jan 02 11:54:52 server2 gunicorn[45000]:     ^
Jan 02 11:54:52 server2 gunicorn[45000]: TabError: inconsistent use of tabs and spaces in indentation

After some testing with the syntax i now use the following code and it seems to work (create a new device with the same serial fails and editing works again).

from extras.validators import CustomValidator

class MyValidator(CustomValidator):
    def validate(self, device):
        from dcim.models import Device
        existing_devices = Device.objects.exclude(pk=device.pk).filter(serial__iexact=device.serial).exists()
        if existing_devices:
            self.fail("Device Serial already in database.", field="serial")

CUSTOM_VALIDATORS = {
    'dcim.device': (
        MyValidator(),
    )
}
